File: Repeated DNA Sequences.java
// https://leetcode.com/problems/repeated-dna-sequences/

class Solution {
    
    public String buildString(char[] letters, int startIdx, int endIdx) {
        
        // If endIdx is bigger than array's length, exit function.
        // If not, an out-of-bound error will occur
        if (endIdx > letters.length) {return null;}
        
        // Create a mutable string
        StringBuffer tmpString = new StringBuffer();
        
        // Cycle array
        for (int i=startIdx; i<endIdx; i++) {
            tmpString.append(letters[i]);
        }
        return tmpString.toString();
    }
    
    public List<String> findRepeatedDnaSequences(String s) {
        
        // If string is less than 10 characters long, return an empty list
        if (s.length() < 10) {return new ArrayList();}
        
        // Create a hashmap that holds string and occurrences as key and value, respectively.
        HashMap<String, Integer> log = new HashMap();
        
        // Cycle string
        char[] letters = s.toCharArray();
        for (int i=0; i<letters.length; i++) {
            
            // Create a temporary string
            String tmpString = buildString(letters, i, i+10);
            
            // If function returned null, then there aren't any more 10 letter sequences
            if (tmpString == null) {break;}
            
            // Update key or add new key to hash map
            if (log.containsKey(tmpString)) {
                log.put(tmpString, log.get(tmpString) + 1);
            }
            else {
                log.put(tmpString, 1);
            }
        }
        
        // Create an array list
        ArrayList<String> ans = new ArrayList();
        
        // Iterate hash map's keys
        for (String curKey: log.keySet()) {
            
            // If key's value is more than 1, add it to array list
            if (log.get(curKey) > 1) {ans.add(curKey);}
        }
        return ans;
    }
}

File: Repeated DNA Sequences.py
# https://leetcode.com/problems/repeated-dna-sequences/

class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        
        # If string has less than 10 characters, exit function
        if len(s) < 10:
            return
        
        # Create a dictionary
        sequenceLog = dict()
        
        # Cycle string
        for i in range(0, len(s)):
            
            # Create substring
            subStr = s[i:i+10]
            
            # If key already exist, update value. If not, add it to dictionary
            if subStr in sequenceLog.keys():
                sequenceLog[subStr] += 1
            else:
                sequenceLog[subStr] = 1
                
        # Cycle dictionary
        ans = list()
        for k, v in sequenceLog.items():
            
            # If substring occurred more than once in s, add it to ans
            if v > 1:
                ans.append(k)
                
        return ans
